realm/realm-library/src/main/java/io/realm/RealmResults.java

```diff
@@ -576,22 +576,15 @@ public double average(String fieldName) {
     public RealmResults<E> distinct(String fieldName) {
         realm.checkIfValid();
         long columnIndex = getColumnIndex(fieldName);
-        TableOrView tableOrView = getTable();
 
-        TableView tableView;
+        TableOrView tableOrView = getTable();
         if (tableOrView instanceof Table) {
-            tableView = ((Table) tableOrView).getDistinctView(columnIndex);
-        } else {
-            tableView = ((TableView) tableOrView).getTable().getDistinctView(columnIndex);
-        }
-
-        RealmResults<E> realmResults;
-        if (realm instanceof DynamicRealm) {
-            realmResults =  (RealmResults<E>) RealmResults.createFromDynamicTableOrView(realm, tableView, className);
+            TableView tableView = ((Table) tableOrView).getDistinctView(columnIndex);
+            this.table = tableView;
```

Comment:
@cmelchior I left it for the compiler to work on it ;) 
It's just that I'm not a big fan of assigning a new instance to a field right away.